### PR TITLE
enumerables

### PR DESCRIPTION
**Pull request by Zibilyonik**
Created requested enumerable methods without using the pre-existing methods.

### CODE REVIEW

**Review by ElijahTobs — CHANGES_REQUESTED** [review]
# Status: Changes Required :recycle: 

## Hi @Zibilyonik :wave: 

You've done well so far :+1:  and I commend you for that :clap: 
However, there are still some changes to be made before this milestone can be approved.

## REQUIRED

* [x] Your project should have a descriptive README.md ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/descriptive_readme.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))
* [x] Rename your enumerable file appropriately.

#### `my_each_with_index`
* [x] When a block is given and self is a hash, the method is expected to behave like the `each_with_index` method which calls the given block once for each element in self.

#### `my_all`
- [x] it is expected to return false if the block returns false or nil for range when tested with the `all` method
- [x] when no block or argument is given, it is expected to return true when none of the elements of the array returns false or nil when tested with the `all` method.
- [x] when a class is passed as an argument, returns true if all of the collection is a member of such class (for integer and numeric values)
- [x] when a Regex is passed as an argument, returns false if any of the collection does not match the Regex
- [x] when a Regex is passed as an argument returns true if all of the collection matches the Regex.

The following should also be revisited as they did not pass the test cases.
- `my_any`
- `my_none`
- `my_count`
- `my_inject`
- `multiply_els`

Check this [this page](https://ruby-doc.org/core-2.6.5/Enumerable.html) to better understand the task before you.

### Make the changes and submit for another review.
## Happy Coding :computer: 

_Feel free to reach out to me on slack_ **`@Elijah Ayandokun`**
  **Elijah O. Ayandokun**
[Github](https://github.com/ElijahTobs) | [LinkedIn](https://www.linkedin.com/in/ayandokunelijah/) | [Twitter](https://twitter.com/ElijahTobs)

**Review by jubaan — CHANGES_REQUESTED** [review]
[![Microverse][microverse-shield]][microverse-url]
<div>
<h1 align="center"> 🚨 🛑 🚨 🛑 INVALID 🚨 🛑 🚨 🛑 </h1>

<br />
<p><q>There's never an easy route to the things that matter.</q></p>
<details open="open">
  <summary>Review Summary</summary>
  <ol>
    <li><a href="#mandatory"> 🧙🏼‍♂️ You shall not pass!! </a> (Why my project was marked as invalid?)</li>
    <li><a href="#comments"> 💬 Comments ✍🏼 ✍🏼 </a></li>
    <li><a href="#contact"> 🧔🏼 About your Reviewer </a> (you don't know it... but we are just automated bots 🤖) </li>
    <li><a href="#help"> 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼 </a></li>
  </ol>
</details>
</div>

<hr>
<h3 id="mandatory"> 1. 🧙🏼‍♂️ You shall not pass!!!!!! (Why my project was marked as invalid?) </h3>
<p><b>Be calm and don't alarm yourself.</b> This is a common mistake and doesn't mean that you haven't put the effort or that the project has something wrong. An <b>INVALID</b> project, only means that the project hasn't met the minimum requirements to be reviewed.</p>
<details close="close">
  <summary><b>[Click Here]</b> to know how to fix it</summary><br />
  <p>Before you submit your code again for review, verify that the project complies with this checklist:</p>

- [ ] Has a professional README
- [ ] Has a descriptive and meaningful PR
- [ ] Commit messages are descriptive and meaningful
- [ ] Linters are set up correctly
- [ ] Project is passing all linters check
- [ ] Have a working project (**doesn't have to be perfect**, but should be able to run)

</details>

<h3 id="optional"> 2. 💡 Comments </h3>
<details close="close">
<summary> [click to open] You might get some useful information..</h3></summary>


- [ ] Nothing to add.


</details>

<h3 id="contact"> 3. 🧔🏼 About your Reviewer </h3>
<details close="close">
<summary><b>[click to open]</b> Say "hello world!" on Slack: @jubaan</summary>
<br />
<div align="center">
  <a src="https://www.github.com/jubaan">
    Julio Añoveros
  </a>
  <p>
    Julio Añoveros | @jubaan</br></br>
    <b>Full-stack Web Developer</b> and</br>
    🤖 Code Reviewer at <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a>:rocket:
  </p>
</div>
</details>

<h3 id="contact"> 4. 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼</h3>
<details close="close">
  <summary> <b>[click to open]</b> <b>Feel stuck?</b> </summary>
  <br>
  <p> First breath... just do it. Take 15-20 min away from the computer, stretch, meditate, do some exercise, drink water, or do whatever is necessary to distract your mind from the obstacle.</p>
  <p><b>Sometimes you just need to get away to see things with a new perspective.</b></p>
</details>

<details close="close">
  <summary> <b>[click to open]</b> <b>Need help?</b> </summary>
  <br>
  <p> Send me a DM in <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> Students Slack, you'll find me as @jubaan. If I'm not available there are other Code Reviewers you can contact. They'll help as well.</p>
  <p>Remember that <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> has other channels where you can get an answer to your questions:</p>
  <ol>
    <li>The <b>students channel 🧑‍🎓 </b> in slack,</li>
    <li>your <b>teammates 👥 </b>,</li>
    <li>your <b>mentor 👨🏼 👩🏼 </b>, and</li>
    <li><b>Student Success 🏫 </b>.</li>
  </ol>
  <p><b> DON'T hesitate to ask! </b></p>
</details>
<p>If all of the above doesn't help... get a <a href="https://rubberduckdebugging.com/">rubber duck</a> ... it would be great to own one 😄. </p>

<h1 align="center"> 🚨 🛑 🚨 🛑 INVALID 🚨 🛑 🚨 🛑 </h1>

[microverse-shield]: https://img.shields.io/badge/-microverse-%236F23FF?style=for-the-badge
[microverse-url]: https://www.microverse.org/?grsf=kfm2ej
[bootstrap-shield]: https://img.shields.io/badge/Bootstrap-white?style=for-the-badge&logo=bootstrap
[bootstrap-url]: https://www.microverse.org/?grsf=kfm2ej

File: enumberable_methods.rb [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    end
+    fin
+  end
+
+  def multiply_els
+    my_inject(1, :*)
+  end
+end
+
+# Arr = [20, 50, 120, 600, 21]
```
Comment (cyonii) [inline enumberable_methods.rb]: Kindly remove all of these comments as they make your code messy ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

**Review by cyonii — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️✍️

Great effort so far, thank you for implementing the previously required changes.

## To Improve
18 of the test cases are still failing, please take a look below and fix the issues.

* [x] Please rename your `enumberable_methods.rb` file to `enumerables.rb` as this is important when running tests against your code.
* [ ] Some of your tests are failing, please fix these methods so that they produce exactly the same result as ruby enumerable methods. You can reference this  [documentation](enumerable) and use it's test cases to make sure your methods are producing the expected results.

### This summary should help you understand what you need to work on to fix the issues.

- `#my_each_with_index` - when a block is given when self is an array, it should behave like `#each_with_index` and call the given block once for each element in self
- `#my_each_with_index` - when a block is given when self is a range, it should return the range itself after calling the given block once for each element in self
- `#my_each_with_index` - when a block is given when self is a hash, it should behave like `#each_with_index` and call the given block once for each element in self
- `#my_all?` - it should return false if the block returns false or nil
- `#my_all?` - when a class is passed as an argument, it should return true if all of the collection is a member of such class
- `#my_all?` - when a class is passed as an argument, it should return true if all of the collection is a member of such class
- `#my_all?` - when a Regex is passed as an argument, it should return true if all of the collection matches the Regex
- `#my_any?` - it should return false if the block ever returns a value false or nil
- `#my_any?` - when a class is passed as an argument, it should return true if at least one of the collection is a member of such class
- `#my_any?` - when a Regex is passed as an argument, it should return true if any of the collection matches the Regex
- `#my_none?` - it should return false if the block ever returns true for all elements
- `#my_none?` - when a Regex is passed as an argument, it should return false if any of the collection matches the Regex
- `#my_count` - it should return the number of items in enum through enumeration
- `#my_inject` - raises a "LocalJumpError" when no block or argument is given
- `#my_inject` - searches for the longest word in an array of strings
- `#my_inject` - when a block is given without an initial value, it should combine all elements of enum by applying a binary operation, specified by a block
- `#my_inject` - when a symbol is specified without an initial value, it should combine each element of the collection by applying the symbol as a named method
- `#multiply_els` - it should accept an array as an argument and multiplies all the elements of the array together using `#my_inject`

### Nice to have (OPTIONAL)
These are optional suggestions, which means you are not obliged to implement them
- [ ] [OPTIONAL] You should not disable so much Rubocop guides as they guide you to write more standard/professional code.


### You can find me on slack 👉 [CY Kalu](https://microverse-students.slack.com/team/U01494U96FR) if you need further assistance or clarification. I'm always available and will be glad to help.

**Comment by Zibilyonik** [comment]
> 
> # Status: Changes Required ♻️✍️
> 
> Great effort so far, thank you for implementing the previously required changes.
> ## To Improve
> 
> 18 of the test cases are still failing, please take a look below and fix the issues.
> 
>     * [x]  Please rename your `enumberable_methods.rb` file to `enumerables.rb` as this is important when running tests against your code.
> 
>     * [ ]  Some of your tests are failing, please fix these methods so that they produce exactly the same result as ruby enumerable methods. You can reference this  [documentation](enumerable) and use it's test cases to make sure your methods are producing the expected results.
> 
> 
> ### This summary should help you understand what you need to work on to fix the issues.
> 
>     * `#my_each_with_index` - when a block is given when self is an array, it should behave like `#each_with_index` and call the given block once for each element in self
> 
>     * `#my_each_with_index` - when a block is given when self is a range, it should return the range itself after calling the given block once for each element in self
> 
>     * `#my_each_with_index` - when a block is given when self is a hash, it should behave like `#each_with_index` and call the given block once for each element in self
> 
>     * `#my_all?` - it should return false if the block returns false or nil
> 
>     * `#my_all?` - when a class is passed as an argument, it should return true if all of the collection is a member of such class
> 
>     * `#my_all?` - when a class is passed as an argument, it should return true if all of the collection is a member of such class
> 
>     * `#my_all?` - when a Regex is passed as an argument, it should return true if all of the collection matches the Regex
> 
>     * `#my_any?` - it should return false if the block ever returns a value false or nil
> 
>     * `#my_any?` - when a class is passed as an argument, it should return true if at least one of the collection is a member of such class
> 
>     * `#my_any?` - when a Regex is passed as an argument, it should return true if any of the collection matches the Regex
> 
>     * `#my_none?` - it should return false if the block ever returns true for all elements
> 
>     * `#my_none?` - when a Regex is passed as an argument, it should return false if any of the collection matches the Regex
> 
>     * `#my_count` - it should return the number of items in enum through enumeration
> 
>     * `#my_inject` - raises a "LocalJumpError" when no block or argument is given
> 
>     * `#my_inject` - searches for the longest word in an array of strings
> 
>     * `#my_inject` - when a block is given without an initial value, it should combine all elements of enum by applying a binary operation, specified by a block
> 
>     * `#my_inject` - when a symbol is specified without an initial value, it should combine each element of the collection by applying the symbol as a named method
> 
>     * `#multiply_els` - it should accept an array as an argument and multiplies all the elements of the array together using `#my_inject`
> 
> 
> ### Nice to have (OPTIONAL)
> 
> These are optional suggestions, which means you are not obliged to implement them
> 
>     * [ ]  [OPTIONAL] You should not disable so much Rubocop guides as they guide you to write more standard/professional code.
> 
> 
> ### You can find me on slack 👉 [CY Kalu](https://microverse-students.slack.com/team/U01494U96FR) if you need further assistance or clarification. I'm always available and will be glad to help.

The documentation link you've provided is broken, please resend the link.

**Review by sinansevgi — APPROVED** [review]

**Comment by sinansevgi** [comment]
# Status: Changes Required ♻️

----

Sorry, I mistakenly approved your project. (This is my first review.)
So far you did a great job but your project still requires improvements.

----
The status of your project will get changed to `uncomplete` soon on your dashboard. 

----

You can create another branch and fix these issues while the status of your project getting updated.


- [x] Your `my_each_with_index ` returns the index of the first appearance of an item in the array. For example, if you pass `[4, 4, 4, 4] `.  It will return `0` as an index of each element instead of real indexes of items. You should keep track of indexes instead of using the array.index method.
- [x] When a class is passed as an argument to #my_all? method, it should return true if all of the collection is a member of that class.
- [x] When a regex is passed as an argument to #my_all? method, it should return true if all of the collection matches the Regex.
- [x] When a class is passed as an argument to #my_any? method, it should return true if any of the collection is a member of that class.
- [x] When a Regex is passed as an argument to #my_any? method, it should return true if any of the collection matches the Regex.
- [x] When a Regex is passed as an argument to #my_none? method, it returns false if any of the collection matches the Regex.
- [x] Your #my_count method doesn't work for ranges. It should return the number of items in the range.
- [x] Your #my_inject method should raise a "LocalJumpError" when no block or argument is given.
- [x] Your #multiple_els method should accept an array as an **argument**.  I couldn't pass an argument to yours.

## I am so sorry for the inconvenience again. 

Please reach me If you have any questions via [Slack](https://microverse-students.slack.com/archives/D0175HK6VME) .

**Pull request merged**